### DISCUSSION TITLE
In Regolith using Wayland and Sway where is the best place to put additional Sway config?

### DISCUSSION THREAD

**Discussion by haruska** [regolith-linux/regolith-desktop#950] [Q&A] [answered]
There are just a couple minor issues preventing me from moving from X11 to Wayland in Regolith. One of them is remapping Caps Lock to Control. The Sway docs recommend adding to 

`~/.config/sway/config`:
```
input "type:keyboard" {
	xkb_options caps:ctrl_modifier
}
```

but this has no effect. I noticed that Regolith 3 stores kanshi config in `~/.config/regolith3/kanshi` instead of the default `~/.config/kanshi` directory. I attempted to move the sway config into `~/.config/regolith3/sway/config` but starting a new session was kind of hosed at that point and I reverted.

Caps Lock is correctly remapped in a Regolith on X11 session using a more traditional xkboptions setup.

Other things tried:
* /etc/default/keyboard is configured to `XKBOPTIONS="ctrl:nocaps"`
* running setxkbmap -option ctrl:nocaps manually

Where should sway config reside? Is there a more straightforward way to remap Caps Lock to Control in Regolith3/Wayland/Sway?

**Comment by dashmage** [comment]
I went through your same situation with remapping CapsLock on Wayland. I used to use `xcape` which only works on X11. Here's my current setup on Wayland for mapping CapsLock to Ctrl while pressing down and using with another key and escape if you tap it. 

I'm using the [caps2esc](https://gitlab.com/interception/linux/plugins/caps2esc) package in order to do this.

Installation of `caps2esc` on ubuntu is quite easy adding the ppa for the apt package.

```sh
sudo add-apt-repository ppa:deafmute/interception
sudo apt install interception-caps2esc
```

Add this yaml file which would then be executed by `udevmon` as a systemd service.
```
$ cat /etc/udevmon.yaml

- JOB: interception -g $DEVNODE | caps2esc | uinput -d $DEVNODE
  DEVICE:
    EVENTS:
      EV_KEY: [KEY_CAPSLOCK, KEY_ESC]
```

`udevmon` comes from the [interception tools](https://gitlab.com/interception/linux/tools) package which is installed as a dependency of `cap2esc`. You can read more about how this works on the README for interception tools.

Create the systemd unit file which manages the interception service.
```
$ cat /etc/systemd/system/udevmon.service

 [Unit]
 Description=udevmon
 Wants=systemd-udev-settle.service
 After=systemd-udev-settle.service

 [Service]
 ExecStart=/usr/bin/nice -n -20 /usr/bin/udevmon -c /etc/udevmon.yaml

 [Install]
 WantedBy=multi-user.target
```

Start the service now and also have it start after boot.
`sudo systemctl enable --now udevmon`

Full credit to [this answer](https://askubuntu.com/a/979606) on AskUbuntu which I referred for the systemd service setup.

**Comment by haruska** [comment]
I would think modifying the sway config in regolith somehow would be more straightforward than installing a udevmon daemon, no? Just unsure where the config should go.

**Reply by dashmage** [reply]
Yeah that would be the easier, more straightforward way I agree. Sorry I couldn't be of more help :)

**Reply by haruska** [reply]
I appreciate you laying out your solution. It was a big help, thank you. If nothing else, it shows the hoops folks are going through to make a seemingly trivial config change.

**Reply by nrvale0** [reply]
@haruska If I read this correctly you were able to find a solution to your key remapping (I use Input Remapper, myself. And for the same reason.) but the top-level issue of where to put custom Sway configs remains unanswered?

If I have that right, I am having the same issue. The sway canonical docs specify a location outside of the ~/.config/regolith3 tree where i3 configs live. I've scoured the regolith3 configs and I'm not totally sure how to proceed.

The config file linked by kgilmer has explicit include statements to paths consistent with the i3 config but that then implies different default behaviors between the two wms. Meaning that regolith has some cleverness to automatically find the i3 configs in ~/.config/regolitth3 however it does not do similar for sway.

**Reply by haruska** [reply]
@nrvale0 You can add sway config at `~/.config/regolith3/sway/config.d/*`. 

To set caps lock to control for example:
```shell
> cat ~/.config/regolith3/sway/config.d/caps_config 
input "type:keyboard" {
	xkb_options caps:ctrl_modifier
}
```

**Comment by kgilmer** [comment, marked as answer]
Hi @haruska ,

This is a documentation issue.  See this line in the root sway config: https://github.com/regolith-linux/regolith-wm-config/blob/main/etc/regolith/sway/config#L166

**Reply by haruska** [reply]
Thanks @kgilmer . For other folks that find this question, to remap caps lock to control in regolith (wayland/sway):

```shell
> cat ~/.config/regolith3/sway/config.d/caps_config 
input "type:keyboard" {
	xkb_options caps:ctrl_modifier
}
```
